src/style/style.js

```diff
@@ -1447,6 +1449,12 @@ class Style extends Evented {
         fog.updateTransitions(parameters);
     }
 
+    _updateMarkersOpacity() {
```

Comment:
this shouldn't be necessary if the markers have their own timers for updating opacity

Reply:
They already have a timer to update their opacity [here](https://github.com/mapbox/mapbox-gl-js/pull/10516/files/e89929b826a19f4400410332615534682047ab73#diff-12738a6ec6570cfc78a8c4e09bec0ff1f9f8762b3372ff9dd80c4fdc9436f24cR528). But this timer is only ever trigger on a [few events](https://github.com/mapbox/mapbox-gl-js/blob/main/src/ui/marker.js#L250) or when mutating the marker.

The reason it is necessary in this case is when we do modifications to terrain or fog or make modification to their properties, removing that section of code, leaves markers in a state where they do not have corrected opacity.

```diff
diff --git a/src/style/style.js b/src/style/style.js
index 3f32c2f2e..ef316e5e3 100644
--- a/src/style/style.js
+++ b/src/style/style.js
@@ -1450,9 +1450,9 @@ class Style extends Evented {
     }
 
     _updateMarkersOpacity() {
-        for (const marker of this.map._markers) {
-            marker._evaluateOpacity();
-        }
+        // for (const marker of this.map._markers) {
+        //     marker._evaluateOpacity();
+        // }
     }
 
     getFog() {
(END)
```
Results in the following (broken in main for the same reasons):

https://user-images.githubusercontent.com/7061573/113035987-95184280-9148-11eb-8476-59d112aefffc.mov

With this branch:

https://user-images.githubusercontent.com/7061573/113035716-51253d80-9148-11eb-942c-6dc53efc26cd.mov

We could create another event for when fog or terrain get updated but that seems cumbersome, so bookkeeping markers seems to be the simplest.